File: src/CsharpClient/Quix.Sdk.Streaming/Models/ParameterValue.cs
using System;
using System.Diagnostics;

namespace Quix.Sdk.Streaming.Models
{
    /// <summary>
    /// Represents a single parameter value of either numeric, string or binary type
    /// </summary>
    [DebuggerDisplay("{" + nameof(Value) + "}")]
    public readonly struct ParameterValue
    {
        private readonly long timestampRawIndex;
        private readonly Parameter parameter;

        internal ParameterValue(long timestampRawIndex, Parameter parameter)
        {
            this.timestampRawIndex = timestampRawIndex;
            this.parameter = parameter;
        }

        /// <summary>
        /// Gets the Parameter Id of the parameter
        /// </summary>
        public readonly string ParameterId => this.parameter.ParameterId;

        /// <summary>
        /// Gets the type of value, which is numeric, string or binary if set, else empty
        /// </summary>
        public readonly ParameterValueType Type
        {
            get
            {
                if (this.parameter.NumericValues != null) return ParameterValueType.Numeric;
                else if (this.parameter.StringValues != null) return ParameterValueType.String;
                else if (this.parameter.BinaryValues != null) return ParameterValueType.Binary;
                else return ParameterValueType.Empty;
            }
        }

        /// <summary>
        /// The numeric value of the parameter.
        /// </summary>
        public double? NumericValue
        {
            get
            {
                return this.parameter.NumericValues?[this.timestampRawIndex];
            }
            set
            {
                if (this.parameter.NumericValues == null)
                {
                    throw new InvalidOperationException($"The parameter '{this.ParameterId}' is not of Numeric type.");
                }

                this.parameter.NumericValues[this.timestampRawIndex] = value;
            }
        }

        /// <summary>
        /// The string value of the parameter
        /// </summary>
        public string StringValue
        {
            get
            {
                return this.parameter.StringValues?[this.timestampRawIndex];
            }
            set
            {
                if (this.parameter.StringValues == null)
                {
                    throw new InvalidOperationException($"The parameter '{this.ParameterId}' is not of String type.");
                }

                this.parameter.StringValues[this.timestampRawIndex] = value;
            }
        }

        /// <summary>
        /// The binary value of the parameter
        /// </summary>
        public byte[] BinaryValue
        {
            get
            {
                return this.parameter.BinaryValues?[this.timestampRawIndex];
            }
            set
            {
                if (this.parameter.BinaryValues == null)
                {
                    throw new InvalidOperationException($"The parameter '{this.ParameterId}' is not of Binary type.");
                }

                this.parameter.BinaryValues[this.timestampRawIndex] = value;
            }
        }

        /// <summary>
        /// Gets the underlying value
        /// </summary>
        public readonly object Value
        {
            get
            {
                return (object)this.parameter.NumericValues?[this.timestampRawIndex]
                    ?? (object)this.parameter.StringValues?[this.timestampRawIndex]
                    ?? (object)this.parameter.BinaryValues?[this.timestampRawIndex];
            }
        }

        /// <summary>
        /// Equality comparison of Parameter values
        /// </summary>
        /// <param name="lhs">First Parameter value to compare</param>
        /// <param name="rhs">Second Parameter value to compare</param>
        /// <returns>Whether the values are equal</returns>
        public static bool operator == (ParameterValue lhs, ParameterValue rhs)
        {
            var lhsValue = lhs.Value;
            var rhsValue = rhs.Value;

            if (lhsValue?.GetType() != rhsValue?.GetType())
            {
                return false;
            }

            if (lhsValue is string)
            {
                return ((string) lhsValue) == ((string) rhsValue);
            }

            if (lhsValue is double?)
            {
                return ((double?) lhsValue) == ((double?) rhsValue);
            }

            if (lhsValue is byte[])
            {
                return ((byte[]) lhsValue) == ((byte[]) rhsValue);
            }

            if (lhsValue == null && rhsValue == null)
            {
                return true;
            }

            return false;
        }

        /// <summary>
        /// Negative equality comparison of Parameter values
        /// </summary>
        /// <param name="lhs">First Parameter value to compare</param>
        /// <param name="rhs">Second Parameter value to compare</param>
        /// <returns>Whether the values are not equal</returns>
        public static bool operator != (ParameterValue lhs, ParameterValue rhs)
        {
            return !(lhs == rhs);
        }

        /// <inheritdoc/>
        public readonly override bool Equals(Object obj)
        {
            return obj is ParameterValue c && this == c;
        }

        /// <inheritdoc/>
        public readonly override int GetHashCode()
        {
            unchecked
            {
                var hash = 397;
                hash ^= this.timestampRawIndex.GetHashCode();
                hash ^= this.parameter.GetHashCode();

                return hash;
            }
        }
    }
}

File: src/PythonClient/lib/quixstreaming/models/parametervalue.py
from typing import Union
from enum import Enum
from .. import __importnet
import Quix.Sdk.Streaming


class ParameterValue(object):
    """
    Represents a single parameter value of either string or numeric type
    """

    def __init__(self, _net_object: Quix.Sdk.Streaming.Models.ParameterValue = None):
        """
            Initializes a new instance of ParameterValue.

            Parameters:

            net_object (.net object): The .net object representing ParameterValue.
        """
        if _net_object is None:
            raise Exception("Parameter value should not be invoked without a .net object")
        self.__wrapped = _net_object

    @property
    def numeric_value(self) -> float:
        """Gets the the numeric value of the parameter, if the underlying parameter is of numeric type"""

        return self.__wrapped.NumericValue

    @numeric_value.setter
    def numeric_value(self, value: float):
        """Sets the the numeric value of the parameter and updates the type to numeric"""

        self.__wrapped.NumericValue = value

    @property
    def string_value(self) -> str:
        """Gets the the string value of the parameter, if the underlying parameter is of string type"""

        return self.__wrapped.StringValue

    @string_value.setter
    def string_value(self, value: str):
        """Sets the the string value of the parameter and updates the type to string"""

        self.__wrapped.StringValue = value

    @property
    def binary_value(self) -> bytes:
        """Gets the the binary value of the parameter, if the underlying parameter is of binary type"""
        val = self.__wrapped.BinaryValue
        if val is None:
            return None
        return bytes(val)

    @binary_value.setter
    def binary_value(self, value: Union[bytearray, bytes]):
        """Sets the the binary value of the parameter and updates the type to binary"""

        self.__wrapped.BinaryValue = value

    @property
    def type(self) -> 'ParameterValueType':
        """Gets the type of value, which is numeric or string if set, else empty"""

        return ParameterValueType.convert_from_net(self.__wrapped.Type)

    @property
    def value(self):
        """Gets the underlying value"""

        return self.__wrapped.Value

    def convert_to_net(self):
        return self.__wrapped


class ParameterValueType(Enum):
    Empty = 0,
    Numeric = 1,
    String = 2,
    Binary = 3

    @staticmethod
    def convert_from_net(net_enum: Quix.Sdk.Streaming.Models.ParameterValueType):
        if net_enum == Quix.Sdk.Streaming.Models.ParameterValueType.Empty:
            return ParameterValueType.Empty
        if net_enum == Quix.Sdk.Streaming.Models.ParameterValueType.Numeric:
            return ParameterValueType.Numeric
        if net_enum == Quix.Sdk.Streaming.Models.ParameterValueType.String:
            return ParameterValueType.String
        if net_enum == Quix.Sdk.Streaming.Models.ParameterValueType.Binary:
            return ParameterValueType.Binary
        raise Exception("Invalid parameter value type")

    def convert_to_net(self):
        if self == ParameterValueType.Empty:
            return Quix.Sdk.Streaming.Models.ParameterValueType.Empty
        if self == ParameterValueType.Numeric:
            return Quix.Sdk.Streaming.Models.ParameterValueType.Numeric
        if self == ParameterValueType.String:
            return Quix.Sdk.Streaming.Models.ParameterValueType.String
        if self == ParameterValueType.Binary:
            return Quix.Sdk.Streaming.Models.ParameterValueType.Binary
